core/server/master/src/main/java/alluxio/master/file/meta/MountTable.java

```diff
@@ -104,47 +117,98 @@ public MountTable(UfsManager ufsManager, MountInfo rootMountInfo) {
    * @throws InvalidPathException if an invalid path is encountered
    */
   public void add(Supplier<JournalContext> journalContext, AlluxioURI alluxioUri, AlluxioURI ufsUri,
-      long mountId, MountPOptions options) throws FileAlreadyExistsException, InvalidPathException {
+      long mountId, MountPOptions options) throws FileAlreadyExistsException, InvalidPathException,
+      IOException {
+    try (LockResource r = new LockResource(mWriteLock)) {
+      // validate the Mount operation first, error will be thrown if the operation is invalid
+      validateMountPoint(alluxioUri, ufsUri);
+      add(journalContext, new Pair<>(alluxioUri, ufsUri), mountId, options);
+    }
+  }
+
+  /**
+   * Inserts an entry into the mount table.
+   * Before calling this method, the caller must hold the write lock to the mount table.
+   * The caller must also have validated the mount information to make sure the mount will succeed.
+   * <blockquote><pre>
+   * try (LockResource mountTableLock = new LockResource(mMountTable.getWriteLock()) {
+   *     mMountTable.validateMountPoint(alluxioPath, ufsPath);
+   *     mMountTable.add(alluxioPath, ufsPath);
+   *     ...
+   *  }
+   * </pre></blockquote>
+   * @param journalContext the journal context
+   * @param validatedPair the validated mount entry [alluxioPath, ufsPath]
+   * @param mountId the mount id
+   * @param options the mount options
+   */
+  public void add(Supplier<JournalContext> journalContext,
```

Comment:
I think it's better to rename as `addValidated` to signal that it's caller's resposibility to make sure the path pair has been validated for mounting.

Reply:
I think it cannot be made private. Since `DefaultFileSystemMaster` will call this method. Is there any ways that can make this method only public to `DefaultFileSystemMaster`?